frontend/apps/web/components/RunTImeline/RunTimeline.tsx

```diff
@@ -0,0 +1,457 @@
+              <div className="h-4 w-[1px] rounded-full bg-gray-500 mx-auto" />
+            </div>
+          ))}
+        </div>
+      </div>
+    );
+  }
+}
+
+function convertTimestampToDate(
+  timestamp: { seconds: bigint; nanos: number } | undefined
+): Date {
+  if (!timestamp) return new Date();
+
+  const millisecondsFromSeconds = Number(timestamp.seconds) * 1000;
+  const millisecondsFromNanos = timestamp.nanos / 1_000_000;
+
+  const totalMilliseconds = millisecondsFromSeconds + millisecondsFromNanos;
+
+  return new Date(totalMilliseconds);
+}
+
+function getCloseOrErrorOrCancelDate(task: JobRunEvent): Date {
+  const errorTask = task.tasks.find((item) => item.error);
+  const errorTime = errorTask ? errorTask.eventTime : undefined;
+  const cancelTime = task.tasks.find(
+    (t) => t.type === 'ActivityTaskCancelRequested'
+  )?.eventTime;
+  return errorTime
+    ? convertTimestampToDate(errorTime)
+    : cancelTime
+      ? convertTimestampToDate(cancelTime)
+      : convertTimestampToDate(task.closeTime);
+}
+
+interface SyncLabelProps {
+  task: JobRunEvent;
+}
+
+function SyncLabel(props: SyncLabelProps) {
+  const { task } = props;
+
+  const schemaTable = `${task.metadata?.metadata.value?.schema}.${task.metadata?.metadata.value?.table} `;
+
+  return (
+    <div className="text-white">
+      {task.metadata?.metadata.case == 'syncMetadata' && schemaTable}
+    </div>
+  );
+}
+
+interface ActivityLabelProps {
+  task: JobRunEvent;
+  getTaskStatus: (task: JobRunEvent) => string;
+}
+
+function ActivityLabel({ task, getTaskStatus }: ActivityLabelProps) {
+  const status = getTaskStatus(task);
+
+  const handleStatus = () => {
+    switch (status) {
+      case 'completed':
+        return <CheckCircledIcon className="text-green-500" />;
+      case 'failed':
+        return <CrossCircledIcon className="text-red-500" />;
+      case 'canceled':
+        return <MinusCircledIcon className="text-yellow-500" />;
+      case 'running':
+        return <Spinner />;
+      default:
+        return null;
+    }
+  };
+
+  return (
+    <div className="flex flex-row items-center gap-2">
+      {task.id.toString()}.
+      <TruncatedText text={task.type} />
+      {handleStatus()}
+    </div>
+  );
+}
+
+interface StatusFilterProps {
+  selectedStatuses: string[];
+  onStatusChange: (status: string, checked: boolean) => void;
+}
+
+// would be nice to replace with a multi-select so you don't have to open/close it everytime you want to make a change
+function StatusFilter({ selectedStatuses, onStatusChange }: StatusFilterProps) {
+  return (
+    <DropdownMenu>
+      <DropdownMenuTrigger asChild>
+        <Button variant="outline">
+          {' '}
+          <MixerHorizontalIcon className="mr-2 h-4 w-4" />
+          Status
+        </Button>
+      </DropdownMenuTrigger>
+      <DropdownMenuContent className="w-56">
+        <DropdownMenuCheckboxItem
+          checked={selectedStatuses.includes('running')}
```

Comment:
This is a nit but can improve optimization.

`selectedStatuses` is a string[]. at the top of the return, let's read this into a set.

```
const uniqueSelectedStatuses = new Set(selectedStatuses)
```

Then all of these calls become really simple.

```
checked={uniqueSelectedStatuses.has('running')}
```

Reply:
updated